AUTOMATAFORMAT.md

```diff
@@ -0,0 +1,111 @@
+* **Bitvector**: Propositional variables, syntax is the same as explicit symbols.
+* **Interval**: Intervals of symbols are of the form [a-z]. We might also want complements, [^a-z], but I am not sure that we want to work with complements. We might also want to give abstract intervals of the form [min,a], [a,max], [min,max], or use inf and -inf instead of min and max. This would be particularly needed if we do not have complements and do not know the alphabet range.
+
+We support specifying **epsilons**, by a key-value line `%Epsilon <formula>` where formula has the same syntax and meaning as the formulae used in transitions (depends on the alphabet type and typing scheme).
+
+## AFA and NFA Examples
+```
+@AFA-bits
+q1 q2 & (a1 | q1 | n1)
+n1 q3 | q4
+q1 a2 & q5
+```
+```
+@AFA-explicit
+%States-enum q r s t "(r,s)"
+%Alphabet-auto
+q symbol | other_symbol & ("(r,s)" | r | s)
+```
+```
+@AFA-intervals
+%States-auto
+%Alphabet-utf
+q [`a`-`z`] | [\u{1c}-\u{5c}] & ("(r,s)" | r | s)
+```
+```
+@NFA-bits
+q (x &  y) | z r
+s x & !y t
+```
+```
+@NFA-intervals
+%Alphabet-utf
+q [`a`-`z`]
+```
+## Transducer
+Transducer has named tracks, and it has the key-value line starting with `%Tracks`. We use `x@<lit>` do say that the <lit> belongs to the track `x`. We may also specify their names by type-identifier or enumeration. Example:
+```
+@AFA-intervals
+%States-auto
+%Tracks-auto
+%Alphabet-utf
+q [`a`-`z`]@x | [\u{1c}-\u{5c}]@z & ("(r,s)" | (r & s))
+```
+``` 
+@AFA-intervals
+%Tracks-auto
+%Alphabet-bits
+q a1@x | a2@z & ("(r,s)" | r | s)
+```
+ 
+## Formulas over symbols and transducer track names.
+The format does not specify the syntax of the formulas, it may be SMT or whatever. We  deal with them in the same way as with all the other types of tokens (symbols, states, nodes). We choose a typing scheme, with the implicit marker `f`. The Thing in the typing scheme specification line is `Formulas`.
+For instance, for transducers, one will want to say things like `f"x=y"`--the symbols on track x and y are the same.
+
+``` 
+@AFA-intervals
+%Tracks-enum x y z
+%Alphabet-enum a b c d e f h
+q (a@x | b@z) & f"z=x" ("(r,s)" | (r & s))
+```
+## Aliases 
+It could be good to allow a akey-value line `%Alias bla bli`, which specifies an alias. The parser will replace occurrences of the token `bla` with the string `bli`.
+
+## Attributes
+We want to assign attributes to states, symbols, nodes, transducer tracks, transitions. Here are two options.
+1. An attribute is a string enclosed in between `{` and `}`. If the specification of the attributes inside the braces uses `{` and `}`, then they must apper escaped, that is, as `\{` and `\}`, as well as the `\`, i.e., `\\`. 
+2. To give an attribute to a transition, we would append `%<attribute>` at the line end.  We could give attributes to states symbols etc by writing key-value lines of the form `%<state> <attribute>`. It sounds easier than the previous version, although perhaps less readable.
```

Comment:
We have already discussed on discord that the first option is slightly better. It is similar to graphviz notation.

Reply:
The question is - do we want to use symbols `{` and `}`? The language of the attribute is regular, so we don't need two bracket symbols, we could use, let's say, apostrophes `'` for opening and closing the attribute block as well.

Reply:
but we loose the graphviz similarity... though, we will probably never be fully compatible with graphviz. (the compatibility is only first impression - I have not studied their attribute grammar)

Reply:
The opening and closing  would in fact become messy. For instance, how to parse 'q & s' & t & a 'q & r' & u'? 
Both brackets and apostropehs have this probem. 
We need to use escapting or some implicit mechanism of pairing beginnings and ends.

Reply:
Your example is a syntax error: unclosed `'`. Escaping is necessary.

Reply:
tu je to atribut hrany (&, a2)
```
q1 a1 & {color: red} a2
```

Reply:
tu je to atribut hrany (&, a1)

```
q1 a1 {color: red} & a2
```

Reply:
Attributy budou oznaceny symbolem 't' a mechanismus bude stejny jako u stavu a symbolu, tedy: t'nejake atributy color: red treba'.